### PR TITLE
memoize domain_has_privilege

### PR DESCRIPTION
**Pull request by benrudolph**
`domain_has_privilege` gets called quite a bit with repeat parameters. not complicated queries, but good to reduce total query count

:leopard:

### CODE REVIEW

**Comment by dannyroberts** [comment]
This does assume that we never care about the effects of a domain's privilege changing during the course of a single request. That seems like a reasonable assumption, but if we do see an odd bug about privileges being briefly stale, this is a good place to look.

**Comment by dannyroberts** [comment]
Also, :heart: it

**Comment by benrudolph** [comment]
Hopefully our requests will be so fast this is a non-factor :sweat_smile:

(also test failures are unrelated)

**Comment by dannyroberts** [comment]
(I restarted just the error build)

**Comment by czue** [comment]
unless i'm missing something this is memoizing in the context of of a module level function, so this will actually cache for much longer than the lifespan of a single request. I don't think we can do this.

**Comment by czue** [comment]
(though if we think the concept is likely to introduce significant performance benefits we should cache it, we'll just have to either use `quickcache` and except that permissions changes might take a few seconds, or we will have to manage invalidation manually.

**Comment by czue** [comment]
just saw https://github.com/dimagi/commcare-hq/pull/5864. so it seems like a different solution with comparable behavior is the best plan here given this does seem like it could noticeably help.

**Comment by benrudolph** [comment]
going to close this for now until i figure out a better solution

**Comment by dannyroberts** [comment]
I agree with @czue's comment about `memoized`. I think `quickcache` would be a good fit for this. Outline of quirks of the two (conceptionally similar) decorators:
- `memoized`
  - in-memory only (so doesn't cache across workers)
  - cache dictionary never gets purged, so if it's on a top-level function, it'll create a memory leak.
  - if it's on an instance method of a class, the cache dictionary is saved on the instance, and so it gets marked for garbage collection along with the instance itself.
- `quickcache`
  - uses tiered caching—first in memory and then in a shared cache (redis)
  - stays alive for a set amount of time, rather than for some natural life-cycle (like a request, or an object's liveness)
  - easily configured to vary on any set of a function's parameters as well as attributes thereof—so for example for a function `def method(self, a, b)` you can vary not only on `a` and `b`, but also specific attributes of self, like `self._id`, etc. or `self.domain`

**Comment by benrudolph** [comment]
wow `quickcache` is pretty slick

**Comment by benrudolph** [comment]
I think the `memoized_timeout` for 10 seconds should be plenty. I'd rather have up to date domain permissions than having to wait 5 minutes for the cache to break which is why I changed `timeout` to 10. Let me know if you guys think that makes sense

**Comment by czue** [comment]
+1 to 10 seconds.

**Comment by czue** [comment]
looks good. i wonder if this breaks the build....

**Comment by benrudolph** [comment]
you would've thought that it would at some point. especially after i read this guy: https://docs.djangoproject.com/en/1.5/topics/testing/overview/#other-test-conditions

File: corehq/apps/accounting/utils.py [hunk 1]
```diff
@@ -98,6 +99,7 @@ def get_change_status(from_plan_version, to_plan_version):
     return adjustment_reason, downgraded_privs, upgraded_privs
 
 
+@quickcache(timeout=10)
```
Comment (snopoke) [inline corehq/apps/accounting/utils.py]: Doesn't this need to specify the `vary_on` arguments?

Comment (snopoke) [inline corehq/apps/accounting/utils.py]: https://github.com/dimagi/commcare-hq/pull/5890

Comment (czue) [inline corehq/apps/accounting/utils.py]: does `quickcache` not default to varying on all arguments?

Comment (snopoke) [inline corehq/apps/accounting/utils.py]: I don't think so

**Pull request merged**